src/GHC/Util/Language/Haskell/GHC/ExactPrint/Types.hs

```diff
@@ -0,0 +1,378 @@
+-- Adapted from https://github.com/alanz/ghc-exactprint.git.
```

Comment:
Could we depend on ghc-exactprint? Copying this code wholescale is scary.

Reply:
If we put our mind to it I suppose we could. My thoughts though are running in the other direction. This module is required by the apply-refact algorithm but without having looked at it deeply, really only for the type declarations - nothing substantial. I have been thinking that until proven otherwise, maybe a better thing to do is rewrite the algorithm to use our own types and drop this module. Something to come back to? Maybe an issue?

Reply:
> Could we depend on ghc-exactprint?
Almost certainly at the cost of introducing longer build times - we'd need `ghc-lib` I think. You'll note comparing the code with the original, I carefully avoided that bullet thus far :) To clarify : it might be a good idea to depend on exact-print and apply-refact and get this for free. Don't know right now.

Reply:
Fine, let's go with what we have now, as its not a big problem right now, and revisit it later.

Reply:
✔️